### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Kawsay**
Dont pull this of course. 

The main motivation is to refactor `crawler.rb` by creating `State` objects.

This is still pretty experimental so I'm more than open to suggestions or help.

See `lib/state/alaska.rb` for an example of refactored state

- State

```ruby
	@code = State code
	@url  = State URL
	@driver = Selenium driver requesting parsing
	@file_time = ID use to create file path (Utils.file_time)
	@s = curl -s @url
	@h = Hash of data
	@errors


	:h
		initialize data Hash and feed it with previous data from all.csv

	:previous 
		gather previous data from all.csv into a Struct
		<Struct :tested, :positive, :deaths>

	:crawl_page(driver: @driver, url: @url)
		navigate to the URL and extract the page source code into a file
		also generates unique file name
 
	:wait
		Selenium::WebDriver::Wait.new(timeout: 60)
		Allow replacement of :sleep by :
			wait.until { #some_condition_to_be true }
			wait.until { @driver.find_element(id: 'ember176').text.empty? }

	:doc
		Nokogiri::HTML(@s)

			
	::STATE_CODE
		Hash storing states' abbreviations and matching class
		{
			AK: Alaska,
			AL: Alabama,
			...
		}

	:self.identify(state_code)
		given a String or Symbol, returns the matching class
```

 - State::YourFavouriteState
```ruby
	State::YourFavouriteState < State

	:perform
		must parse URL, gather data into @h and return it
```

 - Utils
```ruby
module Utils
	:self.file_time

	:self.string_to_i(s)

	:self.h_numbers
```

### CODE REVIEW

**Comment by huuep** [comment]
take a look at Rick's last commit. He pulled the TN parser out to a separate file. Would could start with small steps like that

**Comment by Kawsay** [comment]
Indeed, smaller steps looks more appropriated. I'll close this PR in few hours and will submit smaller ones following Rick's way

File: lib/state.rb [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    # VI: VirginIslands, 
+    VT: Vermont, 
+    WA: Washington, 
+    WI: Wisconsin, 
+    WV: WestVirginia, 
+    WY: Wyoming
+  }
+
+  def self.identify(state_code)
+    STATE_CODES[state_code] || STATE_CODES[state_code.to_sym]
```
[the remark this answers is not in the document]
Comment (Kawsay) [reply lib/state.rb:150]: Thanks Oscar, I'll edit it for my next PR. I think I'll add unit tests once I'll have refactored individuals Crawlers, I keep you informed if I need help with it

File: lib/crawler.rb [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+  #    #end
+  #  end
+
+  # main execution loop
+  # 
+  # default is to run all states in automatic mode
+  # or you can specifiy the list of states to run
+  # if auto_flag is false, will prompt you for certain states
+  #
+  def run(crawl_list = [], auto_flag = true, debug_page_flag = false)
```
Comment (jdpahl122) [inline lib/crawler.rb:140]: If this is a refactor, we should try and break this method down. A 150+ line method is insanity

File: lib/crawler.rb [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+
+        if @h_prev[state_code][:tested] == h[:tested] && @h_prev[state_code][:positive] == h[:positive] && @h_prev[state_code][:deaths] == h[:deaths]
+          # no change
+        elsif @h_prev[state_code][:tested] == h[:tested] && @h_prev[state_code][:positive] == h[:positive]
+          puts "only deaths changed for #{state_code}"
+          puts "old h: #{@h_prev[state_code]}"
+          puts "new h: #{h}"
+          unless @auto_flag
+            @driver.navigate.to(state_url) rescue nil
+            byebug
```
Comment (jdpahl122) [inline lib/crawler.rb:220]: debugger should be removed

File: lib/crawler.rb [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+            puts
+          end
+        elsif @h_prev[state_code][:positive] == h[:positive]
+          if h[:tested] 
+            puts "tested different, positives same for #{state_code}"
+            puts "old h: #{@h_prev[state_code]}"
+            puts "new h: #{h}"
+            unless @auto_flag
+              @driver.navigate.to(state_url) rescue nil
+              byebug
```
Comment (jdpahl122) [inline lib/crawler.rb:230]: here as well

File: lib/crawler.rb [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+              puts
+            end
+          else
+            # missing tested in new
+          end
+        elsif !h[:positive]
+          byebug
```
Comment (jdpahl122) [inline lib/crawler.rb:237]: debugger

**Comment by huuep** [comment]
great feedback everyone. I'll close this PR. Let's continue this work in small steps like what was done in https://github.com/coronavirusapi/crawl-and-parse/pull/27

File: lib/crawler.rb (hunk 2 above)
Comment (Kawsay) [reply lib/crawler.rb:140]: It does not aim to be merged, I shared it to show how I started to move `parse_xx` methods into State::Xx instance methods. But I think we should go for Rick's way, as it wouldn't break everything before having refactored every parse methods

**Pull request closed without merging**